VeriFast solution for Challenge 5 (linked_list.rs)

btj opened the pull request:
This PR adds a VeriFast proof that the LinkedList APIs enumerated in Challenge 5 have the properties enumerated in the Challenge.

Note that VeriFast has some [known unsoundnesses](https://github.com/verifast/verifast/blob/master/tests/rust/README.md) and may also have unknown unsoundnesses, since it is a non-foundational tool (unlike e.g. [RefinedRust](https://plv.mpi-sws.org/refinedrust/).)

This PR is based on the solution that I announced originally in the #29 thread; since then, I have resolved some VeriFast unsoundnesses and made some other improvements (such as bringing down the verification time for linked_list.rs significantly). I will be happy to produce a new VeriFast release and either update this PR or submit a new one to use the new VeriFast release if that is desired.

See some more details in the #29 thread.

Note: the VeriFast tool application issue (#213) is still open; it should probably be resolved before this PR is accepted. I'm submitting this PR at this point to inform the creation of the tool PR.

Resolves #29.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval commented:
Hi @btj, may I ask what was the motivation behind the diff? We do ask people to refrain from making changes to the existing code, so it would be nice to understand the motivation behind it. Thanks!

btj commented:
Well, the challenge says "The solution cannot impact the runtime logic of the standard library" so I figured that changes that 
'obviously' don't affect the generated machine code might be OK. Here are the most important changes I made and the motivation:
- Marked `Node::new`, `Node::into_element`, and `pop_front_node` as `unsafe` because for non-`unsafe` functions, VeriFast enforces RustBelt semantic well-typedness and for these non-public functions this would unnecessarily complicate the proof. As a result, I had to move some calls of these functions into `unsafe` blocks as well.
- Replaced a few instances where the code moved a Box's content out of the box with `Box::into_inner`. Examples: [`Node::into_element`](https://github.com/rust-lang/rust/blob/c290e9de32e8ba6a673ef125fde40eadd395d170/library/alloc/src/collections/linked_list.rs#L161), `remove_current`, `ExtractIf::next`. The MIR generated for this code breaks the Box abstraction and would require a lot of special-case logic in VeriFast to handle.
- Made some implicit lifetime parameters explicit because I need to refer to them in the proof (e.g. `pop_front_node`, `iter`, `cursor_front_mut`, `cursor_back_mut`, `extract_if`)
- In `pop_front_node`, replaced

      let node = Box::from_raw_in(node.as_ptr(), &self.alloc);
      self.head = node.next;

  by

      self.head = (*node.as_ptr()).next;
      let node_ = Box::from_raw_in(node.as_ptr(), &self.alloc);

  because the former dereferences a Box; the latter is significantly simpler to reason about (but on closer thought, it might not be hard to add support for this).
- Moved some `alloc.clone()` calls into a separate `clone_allocator` function assumed to be correct because `clone` is a trait function, the call's semantic well-typedness is not sufficient to verify the call sites, and VeriFast does not yet support attaching special-case specs to particular trait implementations.
- Replaced some higher-order patterns (in particular: `Option::map` and `Option::unwrap_or_else`) by first-order equivalents; VeriFast does not yet support reasoning about closures.
- Enclosed some expressions in braces so as to be able to insert ghost commands
- Factored some subexpressions out into `let`s so as to be able to insert ghost commands
- Replaced some calls of [`NonNull::from`](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.from), which VeriFast does not yet support, by `NonNull::new_unchecked`.
- Replaced some calls of `mem::take` by `mem::replace` because VeriFast does not yet support referring to a type's default value.
- In `split_off`, I made the following changes:
  - Replaced the use of `iter()` with manual iteration; this is much easier to reason about.
  - Replaced the `for` loops by `loop` loops; for now, VeriFast supports only `loop` loops.
- In `LinkedList::drop`:
  - Moved struct DropGuard and the accompanying impl block to the toplevel; VeriFast does not yet support local items.
  - Replaced the `while` loop by a `loop` loop.
- In `next`, replaced `&*node.as_ptr()` by `node.as_ptr()`, i.e. the subsequent few lines of code use a raw pointer instead of a shared reference. This is much easier to reason about because of the aliasing restrictions.
- Still in `next`, replaced `self.len -= 1` by `let len = self.len; self.len = len - 1`, so that I could insert a ghost command between the load and the store.
- In `move_next`, replaced `self.current = current.as_ref().next;` by `self.current = (*current.as_ptr()).next;`, again to avoid the complexity of shared references. I made an analogous change in `move_prev`, `remove_current`, and `ExtractIf::next`.
- In `move_prev`, replaced `list.len()` by `list.len`. `len()` is a non-`unsafe` function and semantic well-typedness is not sufficient for this call site.
- I notice now that two `as_list` functions are not present in my solution but are present in `main`. But they are not present in the commit pointed to from the challenge. So apparently `as_list` was added after the challenge was created.
- In `remove_current`, replaced `self.current?` by a `match` so that I could insert ghost code onto the `None` path.
- In `ExtractIf::next`, I moved a FnMut call into a separate function that I assumed to be correct, because VeriFast does not yet support FnMut calls.
- In `assert_covariance`, moved some nested functions to the toplevel because VeriFast does not yet support nested functions.
- At the bottom of the file, made some impl lifetime parameters explicit because VeriFast currently requires an impl block's parameter list to be identical to the struct's parameter list.

Unless I overlooked something, the remaining changes have been eliminated by recent VeriFast improvements so they are no longer relevant. See [the new version of the diff](https://github.com/btj/verify-rust-std/blob/verifast-25-01-linked-list/verifast-proofs/alloc/collections/linked_list.code-changes.diff); this version of the proof was done with VeriFast 25.01, released today.

tautschnig commented:
@btj What is the status of this PR in light of the CI integration having been merged?

btj commented:
@tautschnig To apply VeriFast to a piece of code, it is necessary to modify the code in minor ways, mostly so that ghost commands can be inserted in the right places. In its current state, the present PR requires a manual check that the modified code is equivalent to the original code. I've started work on a refinement checker that automates this check. Once the checker supports the code modifications I made as part of the linked list proof, I will update this PR and resubmit it for review. In the meantime, you can follow my progress [here](https://github.com/verifast/verifast/commits/master/tests/rust).

btj commented:
Here is an overview of the verified functions, with the properties verified and the list of functions from `linked_list.rs` they call. Note: "correctness" does not necessarily mean "full functional correctness"; it means we verified the function against a specification that was sufficient to verify its caller(s).

| Function | Correctness | Semantic well-typedness | Callees |
| --- | --- | --- | --- |
| `Node::new` | ✓ | ✓ | None |
| `Node::into_element` | ✓ | ✓ | None |
| `LinkedList::push_front_node` | ✓ | | None |
| `LinkedList::pop_front_node` | ✓ | ✓ | None |
| `LinkedList::unlink_node` | ✓ | | None |
| `LinkedList::split_off_after_node` | ✓ | | `LinkedList::new_in` |
| `LinkedList::new` | ✓ | ✓ | None |
| `LinkedList::new_in` | ✓ | ✓ | None |
| `LinkedList::iter` | ✓ | ✓ | None |
| `LinkedList::iter_mut` | ✓ | | None |
| `LinkedList::cursor_front_mut` | ✓ | ✓ | None |
| `LinkedList::cursor_back_mut` | ✓ | ✓ | None |
| `LinkedList::len` | ✓ | ✓ | None |
| `LinkedList::clear` (!) | ✓ | ✓ | `LinkedList::drop` |
| `LinkedList::contains` (!) | | (*) | `LinkedList::iter`, `Iter::next` |
| `LinkedList::push_front` | ✓ | ✓ | `LinkedList::push_front_node` |
| `LinkedList::split_off` (!) | ✓ | ✓ | `LinkedList::len`, `LinkedList::new_in`, `LinkedList::iter_mut`, `IterMut::next`, `IterMut::next_back`, `LinkedList::split_off_after_node` |
| `LinkedList::remove` (!) | | (*) | `LinkedList::len`, `LinkedList::cursor_front_mut`, `CursorMut::move_next`, `CursorMut::remove_current`, `LinkedList::cursor_back_mut`, `CursorMut::move_prev` |
| `LinkedList::retain` (!) | | (*) | `LinkedList::retain_mut` |
| `LinkedList::retain_mut` (!) | | (*) | `LinkedList::cursor_front_mut`, `CursorMut::current`,  `CursorMut::remove_current`, `CursorMut::move_next` |
| `LinkedList::extract_if` (!) | ✓ | ✓ | None |
| `LinkedList::drop` | | (*) | `LinkedList::pop_front_node` |
| `Iter::next` | ✓ | ✓ | None |
| `IterMut::next` | ✓ | | None |
| `IterMut::next_back` | ✓ | | None |
| `CursorMut::move_next` | ✓ | ✓ | None |
| `CursorMut::move_prev` | ✓ | ✓ | None |
| `CursorMut::current` | ✓ | ✓ | None |
| `CursorMut::remove_current` | ✓ | ✓ | `LinkedList::unlink_node` |
| `ExtractIf::next` | ✓ | ✓ | `LinkedList::unlink_node` |

(*): Semantic well-typedness follows from the fact that the function has no `unsafe` blocks, does not directly mutate any fields, and calls only semantically well-typed functions.

The functions marked with (!) are the ones mentioned in the challenge.

ia0 commented:
16-bit platform is the realm of embedded where linked lists are most probably not used. For 32-bit though there is wasm32-unknown-unknown (or variants) which is a platform where it is not completely improbable for linked lists to be used.

btj commented:
The proof currently runs with debug assertions enabled. It probably makes more sense to run it with debug assertions disabled. (We could also run it both ways, but that seems to have only marginal benefit; I guess it would catch UB or bad side effects within the debug assertion itself.)

tautschnig commented:
@btj Please add to this PR a commit that marks the challenge as resolved in `doc/src/challenges/0005-linked-list.md` (cf. #224 for how this was done for challenge 9).

btj commented:
These are all comments dated Feb 6. My understanding is that Remi started a review back then, and that by approving the PR now, these old comments were published as well. They have been superseded by the subsequent discussions that we have had on this PR since Feb 6.
